File: src/test/java/info/szadkowski/contact/throttle/counter/TumblingCounterTest.java
package info.szadkowski.contact.throttle.counter;

import org.assertj.core.api.AbstractBooleanAssert;
import org.assertj.core.api.AbstractLongAssert;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.time.Duration;

import static org.assertj.core.api.Assertions.assertThat;

class TumblingCounterTest {
  private TumblingCounter tumblingCounter;

  @Nested
  class ZeroWindow {

    @BeforeEach
    void setUp() {
      tumblingCounter = new TumblingCounter(Duration.ZERO);
    }

    @Test
    void shouldAlwaysCreateNewWindow() {
      assertNewWindowWithTimestamp(0).isTrue();
      assertNewWindowWithTimestamp(1).isTrue();
    }

    @Test
    void shouldAlwaysCountOne() {
      assertCountWithTimeStamp(0).isEqualTo(1);
      assertCountWithTimeStamp(0).isEqualTo(1);
      assertCountWithTimeStamp(1).isEqualTo(1);
    }
  }

  @Nested
  class NonZeroWindow {

    @BeforeEach
    void setUp() {
      tumblingCounter = new TumblingCounter(Duration.ofMillis(10));
    }

    @Test
    void shouldCreateNewWindowBeforeFirstCall() {
      assertNewWindowWithTimestamp(0).isTrue();
      assertNewWindowWithTimestamp(10).isTrue();
    }

    @Nested
    class FirstZeroTime {

      @BeforeEach
      void setUp() {
        tumblingCounter.count(0L);
      }

      @Test
      void shouldCreateNewWindow() {
        assertNewWindowWithTimestamp(10).isTrue();
      }

      @Test
      void shouldNotCreateNewWindow() {
        assertNewWindowWithTimestamp(0).isFalse();
        assertNewWindowWithTimestamp(1).isFalse();
        assertNewWindowWithTimestamp(9).isFalse();
      }

      @Test
      void shouldReturnOneInNewWindow() {
        assertCountWithTimeStamp(10).isEqualTo(1);
      }

      @Test
      void shouldReturnTwoInSameWindow() {
        assertCountWithTimeStamp(9).isEqualTo(2);
      }

      @Test
      void shouldCountInPast() {
        assertCountWithTimeStamp(10).isEqualTo(1);
        assertCountWithTimeStamp(9).isEqualTo(2);
      }
    }

    @Nested
    class FirstNonZeroTime {

      @BeforeEach
      void setUp() {
        tumblingCounter.count(5);
      }

      @Test
      void shouldCreateNewWindow() {
        assertNewWindowWithTimestamp(15).isTrue();
      }

      @Test
      void shouldNotCreateNewWindow() {
        assertNewWindowWithTimestamp(5).isFalse();
        assertNewWindowWithTimestamp(6).isFalse();
        assertNewWindowWithTimestamp(14).isFalse();
      }

      @Test
      void shouldReturnOneInNewWindow() {
        assertCountWithTimeStamp(15).isEqualTo(1);
      }

      @Test
      void shouldReturnTwoInSameWindow() {
        assertCountWithTimeStamp(14).isEqualTo(2);
      }
    }
  }

  @Test
  void integration() {
    tumblingCounter = new TumblingCounter(Duration.ofMillis(10));

    assertCountWithTimeStamp(1).isEqualTo(1);
    assertCountWithTimeStamp(2).isEqualTo(2);
    assertCountWithTimeStamp(4).isEqualTo(3);
    assertCountWithTimeStamp(10).isEqualTo(4);
    assertCountWithTimeStamp(11).isEqualTo(1);
    assertCountWithTimeStamp(10).isEqualTo(2);
    assertCountWithTimeStamp(11).isEqualTo(3);
    assertCountWithTimeStamp(20).isEqualTo(4);
    assertCountWithTimeStamp(22).isEqualTo(1);
  }

  private AbstractLongAssert<?> assertCountWithTimeStamp(long timeStampMillis) {
    return assertThat(tumblingCounter.count(timeStampMillis));
  }

  private AbstractBooleanAssert<?> assertNewWindowWithTimestamp(long timeStampMillis) {
    return assertThat(tumblingCounter.isNewWindow(timeStampMillis));
  }
}

File: src/test/kotlin/info/szadkowski/contact/throttle/counter/TumblingCounterTest.kt
package info.szadkowski.contact.throttle.counter

import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import strikt.api.DescribeableBuilder
import strikt.api.expectThat
import strikt.assertions.isEqualTo
import strikt.assertions.isFalse
import strikt.assertions.isTrue
import java.time.Duration

class TumblingCounterTest {
    lateinit var tumblingCounter: TumblingCounter

    @Nested
    inner class ZeroWindow {

        @BeforeEach
        fun setUp() {
            tumblingCounter = TumblingCounter(Duration.ZERO)
        }

        @Test
        fun `Should always create new window`() {
            expectNewWindowWithTimestamp(0).isTrue()
            expectNewWindowWithTimestamp(1).isTrue()
        }

        @Test
        fun `Should always count one`() {
            expectCountWithTimeStamp(0).isEqualTo(1)
            expectCountWithTimeStamp(0).isEqualTo(1)
            expectCountWithTimeStamp(1).isEqualTo(1)
        }
    }

    @Nested
    inner class NonZeroWindow {

        @BeforeEach
        fun setUp() {
            tumblingCounter = TumblingCounter(Duration.ofMillis(10))
        }

        @Test
        fun `Should create new window before first call`() {
            expectNewWindowWithTimestamp(0).isTrue()
            expectNewWindowWithTimestamp(10).isTrue()
        }

        @Nested
        inner class FirstZeroTime {

            @BeforeEach
            fun setUp() {
                tumblingCounter.count(0L)
            }

            @Test
            fun `Should create new window`() {
                expectNewWindowWithTimestamp(10).isTrue()
            }

            @Test
            fun `Should not create new window`() {
                expectNewWindowWithTimestamp(0).isFalse()
                expectNewWindowWithTimestamp(1).isFalse()
                expectNewWindowWithTimestamp(9).isFalse()
            }

            @Test
            fun `Should return one in new window`() {
                expectCountWithTimeStamp(10).isEqualTo(1)
            }

            @Test
            fun `Should return two in same window`() {
                expectCountWithTimeStamp(9).isEqualTo(2)
            }

            @Test
            fun `Should count in past`() {
                expectCountWithTimeStamp(10).isEqualTo(1)
                expectCountWithTimeStamp(9).isEqualTo(2)
            }
        }

        @Nested
        inner class FirstNonZeroTime {

            @BeforeEach
            fun setUp() {
                tumblingCounter.count(5)
            }

            @Test
            fun `Should create new window`() {
                expectNewWindowWithTimestamp(15).isTrue()
            }

            @Test
            fun `Should not create new window`() {
                expectNewWindowWithTimestamp(5).isFalse()
                expectNewWindowWithTimestamp(6).isFalse()
                expectNewWindowWithTimestamp(14).isFalse()
            }

            @Test
            fun `Should return one in new window`() {
                expectCountWithTimeStamp(15).isEqualTo(1)
            }

            @Test
            fun `Should return two in same window`() {
                expectCountWithTimeStamp(14).isEqualTo(2)
            }
        }
    }

    @Test
    fun integration() {
        tumblingCounter = TumblingCounter(Duration.ofMillis(10))

        expectCountWithTimeStamp(1).isEqualTo(1)
        expectCountWithTimeStamp(2).isEqualTo(2)
        expectCountWithTimeStamp(4).isEqualTo(3)
        expectCountWithTimeStamp(10).isEqualTo(4)
        expectCountWithTimeStamp(11).isEqualTo(1)
        expectCountWithTimeStamp(10).isEqualTo(2)
        expectCountWithTimeStamp(11).isEqualTo(3)
        expectCountWithTimeStamp(20).isEqualTo(4)
        expectCountWithTimeStamp(22).isEqualTo(1)
    }

    private fun expectCountWithTimeStamp(timeStampMillis: Long): DescribeableBuilder<Long> {
        return expectThat(tumblingCounter.count(timeStampMillis))
    }

    private fun expectNewWindowWithTimestamp(timeStampMillis: Long): DescribeableBuilder<Boolean> {
        return expectThat(tumblingCounter.isNewWindow(timeStampMillis))
    }
}
